### PR TITLE
Optimize long queries by removing loop invariants and reordering table order by size

### PR DESCRIPTION
**Pull request by ylavoie**
By removing the loop invariants, making sure that the smallest tables are used first and adding a few indexes we get a tenfold speed improvement on the long queries.
This algorythm is O(n2) on a big table, thus we have to be carefull with it.
In order to reduce the time, I
- moved the computation of accno outside the loop and precomputed it at the beginning of the migration routine;
- same for the truncated dates, they were moved in an augmented acc_trans table and computed once;
- moved the type computation in acc_trans so it can show AR, AP or GL, like we do with transactions;
- the query order was modified to use the smallest tables in the beginning and the biggest at the end;
The query time dropped by an order of magnitude (10x)

### CODE REVIEW

File: sql/upgrade/sl3.0.sql [hunk 1]
```diff
@@ -227,7 +227,7 @@ create or replace function pg_temp.reconciliation__account_list () returns setof
         coa.accno || ' ' || coa.description as name,
         coa.accno, coa.id as id
     FROM account coa
-         JOIN cr_coa_to_account cta ON cta.chart_id = coa.id
+    JOIN cr_coa_to_account cta ON cta.chart_id = coa.id
```
Comment (ehuelsmann) [inline sql/upgrade/sl3.0.sql]: To what extent does this make a performance difference?

File: sql/upgrade/sl3.0.sql [hunk 2]
```diff
@@ -973,22 +994,16 @@ INSERT INTO invoice (id, trans_id, parts_id, description, qty, allocated,
...
                        memo, approved, cleared_on, voucher_id, invoice_id)
- SELECT lsmb_entry_id, acc_trans.trans_id, (select id
-                    from account
-                   where accno = (select accno
-                                    from :slschema.chart
-                                   where chart.id = acc_trans.chart_id)),
-                                    amount, transdate, source,
-        CASE WHEN cleared IS NOT NULL THEN TRUE ELSE FALSE END, fx_transaction,
-        memo, approved, cleared, vr_id, invoice.id
+ SELECT lsmb_entry_id, acc_trans.trans_id,
```
Comment (ehuelsmann) [inline sql/upgrade/sl3.0.sql]: You stopped adding the `lsmb_entry_id` by removing the fragment above. Did you test this?

File: sql/upgrade/sl3.0.sql [hunk 3]
```diff
@@ -1119,25 +1134,16 @@ INSERT INTO cr_report(chart_id, their_total, submitted, end_date, updated, enter
...
-        WITH types AS ( SELECT id,'AP' AS type FROM :slschema.ap
-                  UNION SELECT id,'AR'         FROM :slschema.ar
-                  UNION SELECT id,'GL'         FROM :slschema.gl)
-        SELECT DISTINCT ac.trans_id, types.type
-        FROM :slschema.acc_trans ac
-        JOIN types ON ac.trans_id = types.id
-        ORDER BY ac.trans_id
-    ) n ON n.trans_id = a.trans_id
-    ORDER BY post_date,cr.id,n.type,a.source ASC NULLS LAST,a.amount
+SELECT cr.id::INT, cr.end_date, a.source, a.type, a.cleared::TIMESTAMP, a.amount::NUMERIC, a.transdate AS post_date, a.lsmb_entry_id
```
Comment (ehuelsmann) [inline sql/upgrade/sl3.0.sql]: The equivalence between the old and new chunks needs some explanation in the commit message, because I'm not seeing it yet...

File: sql/upgrade/sl3.0.sql [hunk 4]
```diff
@@ -1054,7 +1069,7 @@ BEGIN
 END;
 $$ LANGUAGE PLPGSQL;
 
-SELECT payment_migrate(p.id, p.trans_id, cast(p.exchangerate as numeric), p.paymentmethod_id)
+SELECT pg_temp.payment_migrate(p.id, p.trans_id, cast(p.exchangerate as numeric), p.paymentmethod_id)
```
Comment (ehuelsmann) [inline sql/upgrade/sl3.0.sql]: Has nothing to do with performance fixes in the migration. But nice catch. I would have put it in its own commit because it's thematically different from the other changes.

File: sql/upgrade/sl3.0.sql (hunk 2 above)
Comment (ylavoie) [reply sql/upgrade/sl3.0.sql]: Restored

File: sql/upgrade/sl3.0.sql (hunk 4 above)
Comment (ylavoie) [reply sql/upgrade/sl3.0.sql]: This specific fix has nothing to do but payment_migrate is the 3rd longuest query, accounting for almost 10% of migration time.

File: sql/upgrade/sl3.0.sql (hunk 3 above)
Comment (ylavoie) [reply sql/upgrade/sl3.0.sql]: This algorythm is O(n2) on a big table, thus we have to be carefull with it.
In order to reduce the time, I
- moved the computation of accno outside the loop and precomputed it at the beginning of the migration routine;
- same for the truncated dates, they were moved in an augmented `acc_trans` table and computed once;
- moved the type computation in `acc_trans` so it can show AR, AP or GL, like we do with `transactions`;
- the query order was modified to use the smallest tables in the beginning and the biggest at the end;
The query time dropped by an order of magnitude (10x)

Comment (ehuelsmann) [reply sql/upgrade/sl3.0.sql]: Thanks for the explanation, but it belongs in the commit message, so it is retained even if we have to move from GitHub to another provider some day. This explanation would have made a great commit message.
